File: nodejs/src/api/usageTracker.ts
/**
 * Token usage tracking and cost calculation.
 *
 * Handles provider-specific usage extraction, cost calculation, and recording.
 * Usage recording is non-blocking — enqueues to a write queue for batched DB writes.
 */

import { encode } from "gpt-tokenizer";
import { recordUsage as dbRecordUsage } from "../db/database.js";
import { recordTokenUsage } from "./rateLimiter.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Usage {
  input_tokens: number;
  output_tokens: number;
}

/** Minimal provider config for remote token-counting API calls. */
export interface ProviderConfig {
  baseUrl?: string;
  apiKey: string;
}

export interface Cost {
  input_cost: number;
  output_cost: number;
}

export interface UsageRecordParams {
  apiKeyId: string;
  userId?: string;
  providerId: string;
  inputTokens: number;
  outputTokens: number;
  inputCost: number;
  outputCost: number;
  model: string;
}

// ---------------------------------------------------------------------------
// Usage extraction
// ---------------------------------------------------------------------------

/**
 * Extract input/output token counts from a provider response.
 *
 * All provider adapters already convert responses to OpenAI format,
 * so we can rely on the standard `usage` field.
 */
export function extractUsage(
  providerType: string,
  responseData: Record<string, any>
): Usage {
  const usage = responseData.usage ?? {};

  // OpenAI format (native or already converted)
  let inputTokens: number = usage.prompt_tokens ?? 0;
  let outputTokens: number = usage.completion_tokens ?? 0;

  // Fallback: some responses may use different key names
  if (!inputTokens) inputTokens = usage.input_tokens ?? 0;
  if (!outputTokens) outputTokens = usage.output_tokens ?? 0;

  return { input_tokens: inputTokens, output_tokens: outputTokens };
}

/**
 * Extract usage from a provider response, falling back to accurate token counting
 * when the provider did not return any usage data (input_tokens and output_tokens both 0).
 *
 * Accuracy depends on providerType:
 * - OpenAI (openai_chat / openai_response): local BPE tokenizer (gpt-tokenizer)
 * - Anthropic: POST /v1/messages/count_tokens API (requires providerConfig)
 * - Google: POST :countTokens API (requires providerConfig)
 * - Any API failure → falls back to CJK-aware heuristic (estimateTokens)
 *
 * @param body             The original request body (used to extract input text).
 * @param providerConfig   Provider connection config (baseUrl, apiKey) for remote counting.
 * @param modelName        Upstream model name (for selecting the correct tokenizer encoding).
 */
export async function extractUsageWithFallback(
  providerType: string,
  responseData: Record<string, any>,
  body?: Record<string, any>,
  providerConfig?: ProviderConfig,
  modelName?: string,
): Promise<Usage> {
  const usage = extractUsage(providerType, responseData);

  // Count input/output independently — some providers return only one of the two
  if (usage.input_tokens === 0 && body) {
    const inputText = extractInputTextFromBody(body);
    if (inputText) {
      usage.input_tokens = await countTokensAccurate(providerType, inputText, providerConfig, modelName);
    }
  }
  if (usage.output_tokens === 0) {
    const outputText = extractOutputTextFromResponse(responseData);
    if (outputText) {
      usage.output_tokens = await countTokensAccurate(providerType, outputText, providerConfig, modelName);
    }
  }

  return usage;
}

// ---------------------------------------------------------------------------
// Token estimation (CJK-aware heuristic, zero dependencies)
// ---------------------------------------------------------------------------

/**
 * Roughly estimate token count from text.
 *
 * Uses a CJK-aware heuristic: CJK characters (including Hiragana, Katakana,
 * and Hangul) are counted at ~1.2 tokens each; all other characters at ~0.25
 * tokens each (4 chars per token).  This is accurate enough for cost tracking
 * when the upstream provider does not return usage data.
 */
export function estimateTokens(text: string | null | undefined): number {
  if (!text) return 0;
  let cjk = 0;
  let other = 0;
  for (let i = 0; i < text.length; i++) {
    const code = text.charCodeAt(i);
    if (
      (code >= 0x4e00 && code <= 0x9fff) || // CJK Unified Ideographs
      (code >= 0x3400 && code <= 0x4dbf) || // CJK Extension A
      (code >= 0xf900 && code <= 0xfaff) || // CJK Compatibility Ideographs
      (code >= 0x3040 && code <= 0x309f) || // Hiragana
      (code >= 0x30a0 && code <= 0x30ff) || // Katakana
      (code >= 0xac00 && code <= 0xd7af)   // Hangul Syllables
    ) {
      cjk++;
    } else {
      other++;
    }
  }
  return Math.ceil(cjk * 1.2 + other / 4);
}

// ---------------------------------------------------------------------------
// Accurate token counting (provider-specific)
// ---------------------------------------------------------------------------

/** Cache for cl100k_base encoding (lazy loaded on first use). */
let _cl100kEncode: ((text: string) => number[]) | null = null;

/**
 * Map an OpenAI model name to its BPE encoding.
 * - o200k_base: gpt-4o, o1, o3, o4, gpt-4.1, gpt-4.5, gpt-5 and newer
 * - cl100k_base: gpt-4 (non-4o), gpt-3.5 and older
 */
function getOpenAIEncoding(model: string): "o200k_base" | "cl100k_base" {
  const m = model.toLowerCase();
  if (
    m.startsWith("gpt-4o") ||
    m.startsWith("gpt-4.1") ||
    m.startsWith("gpt-4.5") ||
    m.startsWith("gpt-5") ||
    m.startsWith("o1") ||
    m.startsWith("o3") ||
    m.startsWith("o4")
  ) {
    return "o200k_base";
  }
  if (m.startsWith("gpt-4") || m.startsWith("gpt-3.5")) {
    return "cl100k_base";
  }
  // Default to o200k_base for unknown modern models
  return "o200k_base";
}

/**
 * Count tokens using OpenAI's BPE tokenizer (local, zero network).
 * Uses gpt-tokenizer with the model-appropriate encoding.
 */
async function countTokensOpenAI(text: string, model: string): Promise<number> {
  const encoding = getOpenAIEncoding(model);
  if (encoding === "o200k_base") {
    // Default import uses o200k_base
    return encode(text).length;
  }
  // Lazy-load cl100k_base encoding (only needed for gpt-4 / gpt-3.5)
  if (!_cl100kEncode) {
    const mod = await import("gpt-tokenizer/encoding/cl100k_base");
    _cl100kEncode = mod.encode;
  }
  return _cl100kEncode(text).length;
}

/**
 * Count tokens via Anthropic's count_tokens API.
 * POST {baseUrl}/v1/messages/count_tokens
 */
async function countTokensAnthropic(
  text: string,
  model: string,
  config: ProviderConfig,
): Promise<number> {
  const baseUrl = (config.baseUrl || "https://api.anthropic.com").replace(/\/+$/, "");
  const resp = await fetch(`${baseUrl}/v1/messages/count_tokens`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": config.apiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model,
      messages: [{ role: "user", content: text }],
      max_tokens: 1,
    }),
    signal: AbortSignal.timeout(10_000),
  });
  if (!resp.ok) throw new Error(`Anthropic count_tokens HTTP ${resp.status}`);
  const data = (await resp.json()) as { input_tokens: number };
  return data.input_tokens;
}

/**
 * Count tokens via Google's countTokens API.
 * POST {baseUrl}/v1beta/models/{model}:countTokens?key={apiKey}
 */
async function countTokensGoogle(
  text: string,
  model: string,
  config: ProviderConfig,
): Promise<number> {
  const baseUrl = (config.baseUrl || "https://generativelanguage.googleapis.com").replace(/\/+$/, "");
  const resp = await fetch(
    `${baseUrl}/v1beta/models/${model}:countTokens?key=${config.apiKey}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: [{ role: "user", parts: [{ text }] }],
      }),
      signal: AbortSignal.timeout(10_000),
    },
  );
  if (!resp.ok) throw new Error(`Google countTokens HTTP ${resp.status}`);
  const data = (await resp.json()) as { totalTokens: number };
  return data.totalTokens;
}

/**
 * Accurately count tokens using the provider-appropriate method.
 *
 * - OpenAI: local BPE tokenizer (gpt-tokenizer) — zero network
 * - Anthropic: POST /v1/messages/count_tokens API
 * - Google: POST :countTokens API
 *
 * Falls back to CJK-aware heuristic (estimateTokens) on any error or
 * when provider config is missing.
 */
export async function countTokensAccurate(
  providerType: string,
  text: string,
  config?: ProviderConfig,
  modelName?: string,
): Promise<number> {
  try {
    if (providerType === "openai_chat" || providerType === "openai_response") {
      if (modelName) return await countTokensOpenAI(text, modelName);
    }
    if (providerType === "anthropic" && config?.apiKey && modelName) {
      return await countTokensAnthropic(text, modelName, config);
    }
    if (providerType === "google" && config?.apiKey && modelName) {
      return await countTokensGoogle(text, modelName, config);
    }
  } catch (err) {
    console.warn(`[token-count] Accurate counting failed, using heuristic: ${err}`);
  }
  return estimateTokens(text);
}

/**
 * Extract input text from a request body.
 *
 * Supports three API formats:
 * - Chat Completions: body.messages[].content (string or multimodal array)
 * - Responses API: body.input (string or array of {role, content})
 * - Responses API: body.instructions (system prompt string)
 */
export function extractInputTextFromBody(body: Record<string, any> | undefined): string {
  if (!body) return "";
  const parts: string[] = [];

  // Chat Completions format: messages[].content
  const messages = body.messages;
  if (Array.isArray(messages)) {
    for (const msg of messages) {
      if (!msg) continue;
      const content = msg.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const part of content) {
          if (!part) continue;
          if (part.type === "text" && typeof part.text === "string") {
            parts.push(part.text);
          } else if (part.type === "tool_result") {
            // Anthropic tool_result: content can be a string or array of text blocks
            const trContent = part.content;
            if (typeof trContent === "string") {
              parts.push(trContent);
            } else if (Array.isArray(trContent)) {
              for (const tp of trContent) {
                if (tp?.type === "text" && typeof tp.text === "string") parts.push(tp.text);
              }
            }
          } else if (part.type === "tool_use") {
            // Anthropic tool_use: name + input are part of conversation context
            if (typeof part.name === "string") parts.push(part.name);
            if (part.input != null) parts.push(JSON.stringify(part.input));
          }
        }
      }
    }
  }

  // Responses API format: input as string
  const input = body.input;
  if (typeof input === "string") {
    parts.push(input);
  } else if (Array.isArray(input)) {
    // Responses API format: input as array of {role, content}
    for (const item of input) {
      if (!item) continue;
      const content = item.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const part of content) {
          if (typeof part === "string") {
            parts.push(part);
          } else if (part?.type === "text" || part?.type === "input_text") {
            if (typeof part.text === "string") parts.push(part.text);
          }
        }
      }
    }
  }

  // Responses API format: instructions (system prompt)
  if (typeof body.instructions === "string") {
    parts.push(body.instructions);
  }

  // Anthropic Messages API format: system prompt
  // Can be a string or an array of content blocks [{type: "text", text: "..."}]
  const system = body.system;
  if (typeof system === "string") {
    parts.push(system);
  } else if (Array.isArray(system)) {
    for (const part of system) {
      if (part?.type === "text" && typeof part.text === "string") {
        parts.push(part.text);
      }
    }
  }

  // Tool / function definitions — providers count these as input tokens
  // but they are not captured by message content extraction above.
  // We serialize the full schema so the tokenizer can approximate the cost.
  const tools = body.tools;
  if (Array.isArray(tools) && tools.length > 0) {
    parts.push(JSON.stringify(tools));
  }
  // Legacy OpenAI function calling (deprecated but still used by some clients)
  const functions = body.functions;
  if (Array.isArray(functions) && functions.length > 0) {
    parts.push(JSON.stringify(functions));
  }

  return parts.join(" ");
}

/**
 * Extract output text from a non-streaming response.
 *
 * Supports two API response formats:
 * - Chat Completions: choices[].message.content + reasoning_content
 * - Responses API: output[].content[] (with type output_text/text)
 */
export function extractOutputTextFromResponse(responseData: Record<string, any>): string {
  const parts: string[] = [];

  // Chat Completions format: choices[].message.content + reasoning_content
  const choices = responseData.choices;
  if (Array.isArray(choices)) {
    for (const choice of choices) {
      const msg = choice?.message;
      if (!msg) continue;
      if (typeof msg.content === "string") parts.push(msg.content);
      if (typeof msg.reasoning_content === "string") parts.push(msg.reasoning_content);
    }
  }

  // Responses API format: output[].content[]
  const outputItems = responseData.output;
  if (Array.isArray(outputItems)) {
    for (const item of outputItems) {
      if (!item) continue;
      const content = item.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const part of content) {
          if (typeof part === "string") {
            parts.push(part);
          } else if (part && typeof part.text === "string") {
            // type can be "output_text", "text", etc.
            parts.push(part.text);
          }
        }
      }
    }
  }

  return parts.join(" ");
}

// ---------------------------------------------------------------------------
// Cost calculation
// ---------------------------------------------------------------------------

/**
 * Calculate cost in USD based on per-token prices.
 *
 * @param inputPrice   USD per 1M input tokens (from model_prices or providers table)
 * @param outputPrice  USD per 1M output tokens (from model_prices or providers table)
 */
export function calculateCost(
  inputPrice: number | null | undefined,
  outputPrice: number | null | undefined,
  inputTokens: number,
  outputTokens: number,
): Cost {
  const inPrice = inputPrice ?? 0;
  const outPrice = outputPrice ?? 0;

  const inputCost = (inputTokens / 1_000_000) * inPrice;
  const outputCost = (outputTokens / 1_000_000) * outPrice;

  return {
    input_cost: roundTo(inputCost, 8),
    output_cost: roundTo(outputCost, 8),
  };
}

// ---------------------------------------------------------------------------
// Usage recording — enqueues to write queue (non-blocking)
// ---------------------------------------------------------------------------

/**
 * Record a usage event — enqueues for batched DB writing.
 * This is non-blocking and returns immediately.
 */
export async function recordUsage(params: UsageRecordParams): Promise<void> {
  const {
    apiKeyId,
    userId,
    providerId,
    inputTokens,
    outputTokens,
    inputCost,
    outputCost,
    model,
  } = params;

  const totalCost = roundTo(inputCost + outputCost, 8);

  console.log(
    `[usage] api_key=${apiKeyId} provider=${providerId} model=${model} ` +
      `in=${inputTokens} out=${outputTokens} cost=$${totalCost.toFixed(8)}`
  );

  // Record token usage for TPM rate limiting (in-memory sliding window)
  if (userId) {
    recordTokenUsage(userId, apiKeyId, inputTokens + outputTokens);
  }

  // Enqueue for batched writing — no DB I/O here
  dbRecordUsage(
    parseInt(apiKeyId, 10),
    parseInt(providerId, 10),
    inputTokens,
    outputTokens,
    inputCost,
    outputCost,
    model
  );
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function roundTo(n: number, decimals: number): number {
  const factor = Math.pow(10, decimals);
  return Math.round(n * factor) / factor;
}

File: python/api/usage_tracker.py
"""
Token usage tracking and cost calculation.

Handles provider-specific usage extraction, cost calculation, and recording.
"""

from __future__ import annotations

import logging
import math
from typing import Any

import httpx

try:
    import tiktoken
    _HAS_TIKTOKEN = True
except ImportError:
    _HAS_TIKTOKEN = False

logger = logging.getLogger(__name__)

# Provider configuration for accurate token counting
ProviderConfig = dict[str, str]

# ---------------------------------------------------------------------------
# Usage extraction
# ---------------------------------------------------------------------------

def extract_usage(
    provider_type: str,
    response_data: dict[str, Any],
) -> dict[str, int]:
    """Extract input/output token counts from a provider response.

    All provider adapters already convert responses to OpenAI format,
    so we can rely on the standard ``usage`` field.
    """
    usage = response_data.get("usage", {})

    # OpenAI format (native or already converted)
    input_tokens = usage.get("prompt_tokens", 0)
    output_tokens = usage.get("completion_tokens", 0)

    # Fallback: some responses may use different key names
    if not input_tokens:
        input_tokens = usage.get("input_tokens", 0)
    if not output_tokens:
        output_tokens = usage.get("output_tokens", 0)

    return {
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
    }


# ---------------------------------------------------------------------------
# Token estimation fallback (when providers don't return usage data)
# ---------------------------------------------------------------------------

def estimate_tokens(text: str) -> int:
    """Estimate token count from text.

    Uses a simple heuristic:
    - CJK characters: ~1.2 tokens per character
    - Latin/ASCII characters: ~0.25 tokens per character (≈4 chars/token)
    """
    if not text:
        return 0

    cjk_count = 0
    other_count = 0
    for ch in text:
        code = ord(ch)
        # CJK Unified Ideographs + common CJK ranges
        if (
            0x4E00 <= code <= 0x9FFF        # CJK Unified Ideographs
            or 0x3400 <= code <= 0x4DBF     # CJK Extension A
            or 0x3040 <= code <= 0x30FF     # Hiragana + Katakana
            or 0xAC00 <= code <= 0xD7AF     # Hangul Syllables
            or 0xFF00 <= code <= 0xFFEF     # Fullwidth Forms
        ):
            cjk_count += 1
        else:
            other_count += 1

    return max(1, math.ceil(cjk_count * 1.2 + other_count * 0.25))


def extract_input_text_from_body(body: dict[str, Any]) -> str:
    """Extract concatenated input text from a request body.

    Supports three API formats:
    - Chat Completions: body.messages[].content (string or multimodal array)
    - Responses API: body.input (string or array of {role, content})
    - Responses API: body.instructions (system prompt string)
    """
    parts: list[str] = []

    # Chat Completions format: messages[].content
    messages = body.get("messages")
    if isinstance(messages, list):
        for msg in messages:
            if not isinstance(msg, dict):
                continue
            content = msg.get("content")
            if isinstance(content, str):
                parts.append(content)
            elif isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and block.get("type") == "text":
                        parts.append(block.get("text", ""))
                    elif isinstance(block, str):
                        parts.append(block)

    # Responses API format: input as string or array of {role, content}
    input_data = body.get("input")
    if isinstance(input_data, str):
        parts.append(input_data)
    elif isinstance(input_data, list):
        for item in input_data:
            if not isinstance(item, dict):
                continue
            content = item.get("content")
            if isinstance(content, str):
                parts.append(content)
            elif isinstance(content, list):
                for block in content:
                    if isinstance(block, str):
                        parts.append(block)
                    elif isinstance(block, dict):
                        if block.get("type") in ("text", "input_text"):
                            parts.append(block.get("text", ""))

    # Responses API format: instructions (system prompt)
    instructions = body.get("instructions")
    if isinstance(instructions, str):
        parts.append(instructions)

    return " ".join(parts)


def extract_output_text_from_response(response_data: dict[str, Any]) -> str:
    """Extract concatenated output text from a response.

    Supports two API response formats:
    - Chat Completions: choices[].message.content + reasoning_content
    - Responses API: output[].content[] (with type output_text/text)
    """
    parts: list[str] = []

    # Chat Completions format: choices[].message.content + reasoning_content
    choices = response_data.get("choices")
    if isinstance(choices, list):
        for choice in choices:
            if not isinstance(choice, dict):
                continue
            message = choice.get("message", {})
            if isinstance(message, dict):
                content = message.get("content")
                if isinstance(content, str) and content:
                    parts.append(content)
                reasoning = message.get("reasoning_content")
                if isinstance(reasoning, str) and reasoning:
                    parts.append(reasoning)

    # Responses API format: output[].content[]
    output_items = response_data.get("output")
    if isinstance(output_items, list):
        for item in output_items:
            if not isinstance(item, dict):
                continue
            content = item.get("content")
            if isinstance(content, str):
                parts.append(content)
            elif isinstance(content, list):
                for block in content:
                    if isinstance(block, str):
                        parts.append(block)
                    elif isinstance(block, dict):
                        text = block.get("text")
                        if isinstance(text, str) and text:
                            # type can be "output_text", "text", etc.
                            parts.append(text)

    return " ".join(parts)


# ---------------------------------------------------------------------------
# Accurate token counting (tiktoken / count_tokens API)
# ---------------------------------------------------------------------------

def _get_openai_encoding(model_name: str):
    """Get tiktoken encoding for an OpenAI model, with fallback to cl100k_base."""
    if not _HAS_TIKTOKEN:
        raise ImportError("tiktoken not installed")
    try:
        return tiktoken.encoding_for_model(model_name)
    except KeyError:
        return tiktoken.get_encoding("cl100k_base")


def count_tokens_openai(text: str, model_name: str | None = None) -> int:
    """Count tokens using tiktoken (OpenAI BPE tokenizer)."""
    if not text or not _HAS_TIKTOKEN:
        return 0
    encoding = _get_openai_encoding(model_name or "gpt-4o")
    return len(encoding.encode(text))


async def count_tokens_anthropic(
    text: str,
    provider_config: ProviderConfig,
    model_name: str | None = None,
) -> int:
    """Count tokens via Anthropic count_tokens API."""
    if not text:
        return 0
    base_url = provider_config.get("base_url", "").rstrip("/")
    api_key = provider_config.get("api_key", "")
    if not base_url or not api_key:
        return 0
    model = model_name or "claude-sonnet-4-20250514"
    url = f"{base_url}/v1/messages/count_tokens"
    headers = {
        "x-api-key": api_key,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
    }
    body = {
        "model": model,
        "max_tokens": 1,
        "messages": [{"role": "user", "content": text}],
    }
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.post(url, json=body, headers=headers)
        resp.raise_for_status()
        data = resp.json()
        return data.get("input_tokens", 0)


async def count_tokens_google(
    text: str,
    provider_config: ProviderConfig,
    model_name: str | None = None,
) -> int:
    """Count tokens via Google countTokens API."""
    if not text:
        return 0
    base_url = provider_config.get("base_url", "").rstrip("/")
    api_key = provider_config.get("api_key", "")
    if not base_url or not api_key:
        return 0
    model = model_name or "gemini-2.0-flash"
    url = f"{base_url}/v1beta/models/{model}:countTokens?key={api_key}"
    body = {"contents": [{"parts": [{"text": text}]}]}
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.post(url, json=body)
        resp.raise_for_status()
        data = resp.json()
        return data.get("totalTokens", 0)


async def count_tokens_accurate(
    provider_type: str,
    text: str,
    provider_config: ProviderConfig | None = None,
    model_name: str | None = None,
) -> int:
    """Dispatch to provider-specific accurate token counter.

    Returns 0 on any error (caller should fall back to estimate_tokens).
    """
    if not text:
        return 0
    try:
        if provider_type in ("openai", "openai_chat", "openai_response"):
            return count_tokens_openai(text, model_name)
        elif provider_type == "anthropic":
            if not provider_config:
                return 0
            return await count_tokens_anthropic(text, provider_config, model_name)
        elif provider_type == "google":
            if not provider_config:
                return 0
            return await count_tokens_google(text, provider_config, model_name)
    except Exception:
        logger.debug("Accurate token counting failed, falling back to heuristic")
    return 0


async def extract_usage_with_fallback(
    provider_type: str,
    response_data: dict[str, Any],
    body: dict[str, Any] | None = None,
    provider_config: ProviderConfig | None = None,
    model_name: str | None = None,
) -> dict[str, int]:
    """Extract usage; if no usage data, count accurately or estimate."""
    usage = extract_usage(provider_type, response_data)

    # Count input/output independently — some providers return only one of the two
    if usage["input_tokens"] == 0 and body:
        input_text = extract_input_text_from_body(body)
        if input_text:
            count = await count_tokens_accurate(provider_type, input_text, provider_config, model_name)
            usage["input_tokens"] = count if count else estimate_tokens(input_text)
    if usage["output_tokens"] == 0:
        output_text = extract_output_text_from_response(response_data)
        if output_text:
            count = await count_tokens_accurate(provider_type, output_text, provider_config, model_name)
            usage["output_tokens"] = count if count else estimate_tokens(output_text)

    return usage


# ---------------------------------------------------------------------------
# Cost calculation
# ---------------------------------------------------------------------------

def calculate_cost(
    input_price: float | None,
    output_price: float | None,
    input_tokens: int,
    output_tokens: int,
) -> dict[str, float]:
    """Calculate cost in USD based on per-token prices from the database.

    Parameters
    ----------
    input_price:
        USD per 1M input tokens (from model_prices or providers table).
    output_price:
        USD per 1M output tokens (from model_prices or providers table).
    input_tokens:
        Number of input (prompt) tokens.
    output_tokens:
        Number of output (completion) tokens.

    Returns
    -------
    {"input_cost": float, "output_cost": float} in USD.
    """
    in_price = input_price or 0.0
    out_price = output_price or 0.0

    input_cost = (input_tokens / 1_000_000.0) * in_price
    output_cost = (output_tokens / 1_000_000.0) * out_price

    return {
        "input_cost": round(input_cost, 8),
        "output_cost": round(output_cost, 8),
    }


# ---------------------------------------------------------------------------
# Usage recording
# ---------------------------------------------------------------------------

async def record_usage(
    api_key_id: int,
    provider_id: int,
    input_tokens: int,
    output_tokens: int,
    input_cost: float,
    output_cost: float,
    model: str,
) -> None:
    """Enqueue a usage event for batch writing to the database.

    The actual DB write happens in a background flush (every 5s or at 100 entries).
    """
    from db.database import enqueue_usage

    total_cost = round(input_cost + output_cost, 8)

    logger.info(
        "Usage: api_key=%s provider=%s model=%s in=%d out=%d cost=$%.8f",
        api_key_id,
        provider_id,
        model,
        input_tokens,
        output_tokens,
        total_cost,
    )

    enqueue_usage(
        api_key_id=api_key_id,
        provider_id=int(provider_id),
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        input_cost=input_cost,
        output_cost=output_cost,
        model=model,
    )
